File: frontend/src/utils/portfolioSummary.ts
import type { PortfolioSummary } from '@/hooks/usePortfolio'

export interface PortfolioSummaryMetrics {
  patrimonio: number
  aportado: number
  lucroTotal: number
  variacaoValor: number
  variacaoPct: number
  ganhoNaoRealizado: number
  ganhoRealizado: number
  rentabilidadePct: number
  rentabilidadeDiariaPct: number | null
  proventos12m: number
  proventosTotal: number
  hasPartialPrices: boolean
  assetsWithoutPrice: string[]
  valuationMode: string | null
  valuationUpdatedAt: string | null
  performanceAsOf: string | null
  proventosAsOf: string | null
  rentabilidadeSource: string | null
  returnIsEstimated: boolean
}

export function formatReferenceDate(value?: string | null): string | null {
  if (!value) return null
  const dateOnly = /^\d{4}-\d{2}-\d{2}$/.test(value)
  const parsed = new Date(dateOnly ? `${value}T12:00:00` : value)
  if (Number.isNaN(parsed.getTime())) return null
  return new Intl.DateTimeFormat('pt-BR', dateOnly
    ? { day: '2-digit', month: '2-digit', year: 'numeric' }
    : { day: '2-digit', month: '2-digit', year: 'numeric', hour: '2-digit', minute: '2-digit' }
  ).format(parsed)
}

/**
 * Mapeia exclusivamente o contrato consolidado usado pelo Resumo.
 * Valores intradiários e performance fechada mantêm referências temporais distintas.
 */
export function mapPortfolioSummaryMetrics(summary: PortfolioSummary): PortfolioSummaryMetrics {
  return {
    patrimonio: summary.total_patrimonio ?? 0,
    aportado: summary.total_investido ?? 0,
    lucroTotal: summary.lucro_total ?? 0,
    variacaoValor: summary.variacao_valor ?? 0,
    variacaoPct: summary.variacao_percentual ?? 0,
    ganhoNaoRealizado: summary.ganho_nao_realizado ?? 0,
    ganhoRealizado: summary.ganho_realizado ?? 0,
    rentabilidadePct: summary.rentabilidade_total ?? 0,
    rentabilidadeDiariaPct: summary.rentabilidade_diaria ?? null,
    proventos12m: summary.dividendos_recebidos_12m ?? 0,
    proventosTotal: summary.total_proventos ?? 0,
    hasPartialPrices: summary.has_partial_prices ?? false,
    assetsWithoutPrice: summary.assets_without_price ?? [],
    valuationMode: summary.valuation_mode ?? null,
    valuationUpdatedAt: summary.valuation_updated_at ?? null,
    performanceAsOf: summary.performance_as_of ?? null,
    proventosAsOf: summary.proventos_as_of ?? null,
    rentabilidadeSource: summary.rentabilidade_source ?? null,
    returnIsEstimated: summary.return_is_estimated ?? false,
  }
}

export function getPortfolioReturnPresentation(
  metrics: PortfolioSummaryMetrics | null,
): { isEstimated: boolean; label: 'Retorno estimado' | 'Rentabilidade (TWR)' } {
  const isEstimated = Boolean(
    metrics?.returnIsEstimated || metrics?.rentabilidadeSource === 'valuation_fallback',
  )
  return {
    isEstimated,
    label: isEstimated ? 'Retorno estimado' : 'Rentabilidade (TWR)',
  }
}

File: backend/app/schemas/portfolio_summary.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class PortfolioSummaryResponse(BaseModel):
    """Contrato público e versionado da página Resumo."""

    model_config = ConfigDict(extra="forbid")

    summary_version: Literal["summary.v2"]

    total_patrimonio: float
    total_investido: float
    lucro_total: float
    variacao_valor: float
    variacao_percentual: float
    ganho_nao_realizado: float
    ganho_realizado: float

    rentabilidade_total: float
    rentabilidade_acumulada: float
    rentabilidade_diaria: float | None
    rentabilidade_source: Literal["snapshot_twr", "valuation_fallback"]

    dividendos_recebidos_12m: float
    total_proventos: float
    proventos_as_of: str
    proventos_source: Literal["received_cash_dividends"]

    has_partial_prices: bool
    assets_without_price: list[str]
    price_assets_total: int = Field(ge=0)
    price_assets_covered: int = Field(ge=0)
    price_coverage_pct: float = Field(ge=0, le=100)

    usd_brl_rate: float
    valuation_mode: Literal["intraday"]
    valuation_updated_at: str | None
    performance_as_of: str | None
    snapshot_id: int | None
    snapshot_date: str | None
    summary_source: Literal[
        "intraday_valuation_with_snapshot_twr",
        "valuation_fallback",
    ]
    return_is_estimated: bool

    is_reconciled: bool | None
    reconciliation: dict[str, Any] | None
